src/symbolic_dimensions.jl

```diff
@@ -35,49 +34,54 @@ to one which uses `Dimensions` as its dimensions (i.e., base SI units)
 `expand_units`.
 """
 struct SymbolicDimensions{R} <: AbstractDimensions{R}
-    _data::SA.SparseVector{R}
-
-    SymbolicDimensions(data::SA.SparseVector) = new{eltype(data)}(data)
-    SymbolicDimensions{_R}(data::SA.SparseVector) where {_R} = new{_R}(data)
+    nzdims::Vector{UInt8}
+    nzvals::Vector{R}
 end
 
 static_fieldnames(::Type{<:SymbolicDimensions}) = ALL_SYMBOLS
-data(d::SymbolicDimensions) = getfield(d, :_data)
-Base.getproperty(d::SymbolicDimensions{R}, s::Symbol) where {R} = data(d)[ALL_MAPPING[s]]
-Base.getindex(d::SymbolicDimensions{R}, k::Symbol) where {R} = getproperty(d, k)
+function Base.getproperty(d::SymbolicDimensions{R}, s::Symbol) where {R}
+    nzdims = getfield(d, :nzdims)
+    i = ALL_MAPPING[s]
+    ii = searchsortedfirst(nzdims, i)
+    (ii <= length(nzdims) && nzdims[ii] == i) ? getfield(d, :nzvals)[ii] : zero(R)
+end
+Base.propertynames(::SymbolicDimensions) = ALL_SYMBOLS
+Base.getindex(d::SymbolicDimensions, k::Symbol) = getproperty(d, k)
 constructor_of(::Type{<:SymbolicDimensions}) = SymbolicDimensions
 
-SymbolicDimensions{R}(d::SymbolicDimensions) where {R} = SymbolicDimensions{R}(data(d))
-(::Type{D})(; kws...) where {D<:SymbolicDimensions} = D(DEFAULT_DIM_BASE_TYPE; kws...)
-(::Type{D})(::Type{R}; kws...) where {R,D<:SymbolicDimensions} =
-    let constructor=constructor_of(D){R}
-        length(kws) == 0 && return constructor(SA.spzeros(R, length(ALL_SYMBOLS)))
-        I = [ALL_MAPPING[s] for s in keys(kws)]
-        V = [tryrationalize(R, v) for v in values(kws)]
-        data = SA.sparsevec(I, V, length(ALL_SYMBOLS))
-        return constructor(data)
+SymbolicDimensions{R}(d::SymbolicDimensions) where {R} = SymbolicDimensions{R}(getfield(d, :nzdims), convert(Vector{R}, getfield(d, :nzvals)))
+SymbolicDimensions(; kws...) = SymbolicDimensions{DEFAULT_DIM_BASE_TYPE}(; kws...)
+function SymbolicDimensions{R}(; kws...) where {R}
+    if isempty(kws)
+        return SymbolicDimensions{R}(Vector{UInt8}(undef, 0), Vector{R}(undef, 0))
     end
+    I = UInt8[ALL_MAPPING[s] for s in keys(kws)]
+    p = sortperm(I)
+    V = R[tryrationalize(R, kws[i]) for i in p]
+    return SymbolicDimensions{R}(permute!(I, p), V)
+end
+(::Type{<:SymbolicDimensions})(::Type{R}; kws...) where {R} = SymbolicDimensions{R}(; kws...)
```

Comment:
I added this for backward compatibility but IMO it should be removed - the standard way to specify the type `R` is to call `SymbolicDimensions{R}` directly. With this function signature, unintuitive and confusing calls such as `SymbolicDimensions{Float32}(Float64)` are possible.

Reply:
> I don't actually remember if SymbolicDimensions{T}(T2; ...) is used anywhere, I don't think it is?

It's supported here:
https://github.com/SymbolicML/DynamicQuantities.jl/blob/87e1eb3d0d4c139ed94a563e2fae3cababc5e512/src/symbolic_dimensions.jl#L52

Similar definitions exist for other types as well:
https://github.com/SymbolicML/DynamicQuantities.jl/blob/87e1eb3d0d4c139ed94a563e2fae3cababc5e512/src/types.jl#L75
https://github.com/SymbolicML/DynamicQuantities.jl/blob/87e1eb3d0d4c139ed94a563e2fae3cababc5e512/src/types.jl#L118-L119

They all allow such inconsistent calls. IMO all these `(::Type{...})` constructors should be removed and replaced with standard constructors such as e.g. `Dimension(...)` and `Dimension{R}(...)`.

Reply:
> They all allow such inconsistent calls.

The `constructor_of` part removes the type, so I think it's safe, right? It would never actually hit `SymbolicDimensions{R}(R2, ...)` unless you explicitly call that (in which case it would return a `SymbolicDimensions{R2}`.

But yeah I do agree this feels a bit dangerous to play with such generic constructors.

> IMO all these `(::Type{...})` constructors should be removed and replaced with standard constructors such as e.g. `Dimension(...)` and `Dimension{R}(...)`.

I set it up this way so you could easily create your own custom dimensions (as a struct-based alternative to the macro-based approach of Unitful):

```julia
struct MyDimensions{R} <:AbstractDimensions{R}
    length::R
    mass::R
end
```

and just start using it, like:

```julia
x1 = Quantity(0.5, MyDimensions, length=2)
x2 = Quantity(0.3, MyDimensions, mass=-1)

x1 * x2 # 0.5*0.3 length=2 mass=-1
```

I could see many applications where one simply never uses some of the physical dimensions, so I wanted to make it very easy to declare custom dimensional spaces.

But I'm open to opinions on this!

Reply:
> It would never actually hit SymbolicDimensions{R}(R2, ...) unless you explicitly call that (in which case it would return a SymbolicDimensions{R2}.

Yes, the implementation is safe. My point was that it's surprising to allow this call at all and that the return type is unexpected: By convention, you would expect that a call of `SomeType{R}(...)` creates an instance of type `SomeType{R}`.

Reply:
Ah, I see what you mean. I agree we should change that. 

I think the most logical behavior would be:

1. Calling `Dimension` and passing a type `R` => Creates `Dimension{R}`.
2. Calling `Dimension{R}` and passing a type `R` => Creates `Dimension{R}`.
2. Calling `Dimension{R}` explicitly but passing a different type `R2` => Error.

Then for quantities:

1. Calling `Quantity` and passing a value of type `T` => Creates `Quantity{T,D}` (default `D`).
2. Calling `Quantity{T}` and passing a value of type `T2` => Creates `Quantity{T,D}` (no error) (default `D`).
2. Calling `Quantity{T}` and passing a value of type `T2` and specifying a dimension type `D` => Creates `Quantity{T,D}`.
3. Calling `Quantity{T,D}` and passing a value of type `T2` => Creates `Quantity{T,D}`.
3. Calling `Quantity{T,D}` and passing a value of type `T2` and specifying dimension type `D` => Creates `Quantity{T,D}`.
4. Calling `Quantity{T,D}` and passing a value of any type, but specifying a different dimension type `D2` => Error.

Wdyt?

Reply:
I think even better than erroring out in some cases would be to not allow such calls in the first place. For instance, for `Dimension` I would just define
```julia
Dimension(; kw...) = Dimension{DEFAULT_BASE_TYPE}(; kw...)
Dimension{R}(; length=zero(R), time=zero(R), ....) where {R} = Dimension{R}(R(length), R(time), ...)
```
without any specific `Type` argument.

Also for `Quantity` `Type` arguments are not needed: One could define
```julia
Quantity(val; kw...) = Quantity{typeof(val),DEFAULT_DIMENSION_TYPE}(val; kw...)
Quantity{T}(val; kw...) where {T} = Quantity{T,DEFAULT_DIMENSION_TYPE}(val; kw...)
Quantity{T,D}(val; kw...) where {T,D<:Dimension} = Quantity(T(val), D(; kw...))
```

If it's more desirable to specify the dimension type, then one could change the order of the type parameters.

For user types, the automatic keyword-based functionality (i.e., the `Quantity{T,<:CustomType}(...)` definitions) could be defined with a weak dependency on ConstructionBase.jl: https://juliaobjects.github.io/ConstructionBase.jl/dev/ If users want to use this convenience feature for their custom dimension types, they could extend `ConstructionBase.constructorof`. In the extension one would define
```julia
function DynamicQuantities.Quantity{T,D}(val; kw...) where {T,D<:AbstractDimension}
    return Quantity(T(val), ConstructionBase.constructorof(D)(; kw...))
end
```

Reply:
I think we could solve this in another PR, and just get this one out of the way first? Since this would be a breaking change anyways, we could get the SparseArrays removal into v0.7 first.

Reply:
Yes, I agree. Such changes are not independent of the SparseArrays PR.